homeassistant/components/harmony/config_flow.py

```diff
@@ -0,0 +1,211 @@
+
+            if "base" not in errors:
+                return await self._async_create_entry_from_valid_input(info, user_input)
+
+        # Return form
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a discovered Harmony device."""
+        _LOGGER.debug("SSDP discovery_info: %s", discovery_info)
+
+        parsed_url = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION])
+        friendly_name = discovery_info[ssdp.ATTR_UPNP_FRIENDLY_NAME]
+
+        # pylint: disable=no-member
+        self.context["title_placeholders"] = {"name": friendly_name}
+
+        self.harmony_config = {
+            CONF_HOST: parsed_url.hostname,
+            CONF_NAME: friendly_name,
+        }
+
+        if self._host_already_configured(self.harmony_config):
+            return self.async_abort(reason="already_configured")
+
+        return await self.async_step_link()
+
+    async def async_step_link(self, user_input=None):
+        """Attempt to link with the Harmony."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, self.harmony_config)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if "base" not in errors:
+                return await self._async_create_entry_from_valid_input(info, user_input)
+
+        return self.async_show_form(
+            step_id="link",
+            errors=errors,
+            description_placeholders={
+                CONF_HOST: self.harmony_config[CONF_NAME],
+                CONF_NAME: self.harmony_config[CONF_HOST],
+            },
+        )
+
+    async def async_step_import(self, user_input):
+        """Handle import."""
+        return await self.async_step_user(user_input)
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return OptionsFlowHandler(config_entry)
+
+    async def _async_create_entry_from_valid_input(self, validated, user_input):
+        """Single path to create the config entry from validated input."""
+        await self.async_set_unique_id(validated[UNIQUE_ID])
+        if self._host_already_configured(validated):
+            return self.async_abort(reason="already_configured")
+        self._abort_if_unique_id_configured()
+        config_entry = self.async_create_entry(
+            title=validated[CONF_NAME],
+            data={CONF_NAME: validated[CONF_NAME], CONF_HOST: validated[CONF_HOST]},
+        )
+        # Options from yaml are preserved
+        options = _options_from_user_input(user_input)
+        if options:
+            config_entry["options"] = options
```

Comment:
Where are these options handled?

Reply:
They are handled in the `OptionsFlowHandler`.  These are only for importing from yaml

Reply:
I don't see where that happens. The `config_entry` here is not an instance of `ConfigEntry`. It's just the result dictionary returned by `async_create_entry`.

Reply:
ConfigEntriesFlowManager.async_finish_flow always sets the options to {} so it makes it impossible to setup the options on import AFAICT.

I could either.

1. Store it in data and then pull them out if they are missing.

2. Modify homeassistant/config_entries.py to allow passing in options
```
diff --git a/homeassistant/config_entries.py b/homeassistant/config_entries.py
index 945bd3865..90df0a566 100644
--- a/homeassistant/config_entries.py
+++ b/homeassistant/config_entries.py
@@ -469,7 +469,7 @@ class ConfigEntriesFlowManager(data_entry_flow.FlowManager):
             domain=result["handler"],
             title=result["title"],
             data=result["data"],
-            options={},
+            options=result.get("options,{}),
             system_options={},
             source=flow.context["source"],
             connection_class=flow.CONNECTION_CLASS,
```

3. Any other ideas?

Reply:
Currently we do 1, ie we store it in config_entry.data and populate options with default values from that.

I agree that this feels like a missing feature of the current config flow api. But that needs some discussion first, eg in an issue in our architecture repo.

Reply:
Thanks. I've got the followup here https://github.com/home-assistant/core/pull/33018

Reply:
Previously Paulus wanted config options to have sane defaults and only allow changing them after setup. There is some material that would motivate changing this direction now